ui/src/app/pages/login/sign-in.tsx

```diff
@@ -127,127 +127,177 @@ interface SignInState {
+      // data in its onComplete callback.
+      profile: this.createEmptyProfile()
+    };
+  }
 
-    createEmptyProfile(): Profile {
-      return {
-        // Note: We abuse the "username" field here by omitting "@domain.org". After
-        // profile creation, this field is populated with the full email address.
-        username: '',
-        dataAccessLevel: DataAccessLevel.Unregistered,
-        givenName: '',
-        familyName: '',
-        contactEmail: '',
-        currentPosition: '',
-        organization: '',
-        areaOfResearch: '',
-        address: {
-          streetAddress1: '',
-          streetAddress2: '',
-          city: '',
-          state: '',
-          country: '',
-          zipCode: '',
+  private createEmptyProfile(): Profile {
+    return {
+      // Note: We abuse the "username" field here by omitting "@domain.org". After
+      // profile creation, this field is populated with the full email address.
+      username: '',
+      dataAccessLevel: DataAccessLevel.Unregistered,
+      givenName: '',
+      familyName: '',
+      contactEmail: '',
+      currentPosition: '',
+      organization: '',
+      areaOfResearch: '',
+      address: {
+        streetAddress1: '',
+        streetAddress2: '',
+        city: '',
+        state: '',
+        country: '',
+        zipCode: '',
+      },
+      institutionalAffiliations: [
+        // We only allow entering a single institutional affiliation from the create account
+        // page, so we pre-fill a single empty entry which will be bound to the AccountCreation
+        // form.
+        {
+          institution: undefined,
+          nonAcademicAffiliation: undefined,
+          role: undefined,
         },
-        institutionalAffiliations: [
-          // We only allow entering a single institutional affiliation from the create account
-          // page, so we pre-fill a single empty entry which will be bound to the AccountCreation
-          // form.
-          {
-            institution: undefined,
-            nonAcademicAffiliation: undefined,
-            role: undefined,
-          },
-        ],
-        demographicSurvey: {},
-        degrees: [] as Degree[],
-      };
+      ],
+      demographicSurvey: {},
+      degrees: [] as Degree[],
+    };
+  }
+
+  componentDidMount() {
+    document.body.style.backgroundColor = colors.light;
+    this.props.onInit();
+  }
+
+  /**
+   * Creates the appropriate set of steps based on the server-side config.
+   *
+   * Made visible for ease of unit-testing.
+   */
+  public getAccountCreationSteps(): Array<SignInStep> {
+    let steps = [
+      SignInStep.LANDING,
+      SignInStep.INVITATION_KEY,
+      SignInStep.TERMS_OF_SERVICE,
+      SignInStep.ACCOUNT_CREATION,
+      SignInStep.DEMOGRAPHIC_SURVEY,
+      SignInStep.SUCCESS_PAGE
+    ];
+
+    const {enableNewAccountCreation, requireInvitationKey} = this.props.serverConfig;
+
+    if (!enableNewAccountCreation) {
```

Comment:
nit: I'd prefer to just see separate lists returned in each of these cases. In this case, we're not going to change what `enableNewAccountCreation` or `requireInvitationKey` means.